### PR TITLE
DBZ-2094 Allow Postgres snapshotter to set streaming start on resume

### PR DESCRIPTION
**Pull request by gunnarmorling**
When a connector resumes after previously streaming and takes a
snapshot, through a new method on the snapshotter interface,
shouldStreamEventsStartingFromSnapshot, can choose whether
to resume streaming from the last known streaming position or the
beginning of the snapshot. This is helpful for snapshotters that
may not want to resnapshot every table in the whitelist/blacklist
but not miss event on the tables that are skipped.

Supersedes https://github.com/debezium/debezium/pull/1519 after rebasing.

https://issues.jboss.org/browse/DBZ-2094

### CODE REVIEW

File: debezium-core/src/main/java/io/debezium/pipeline/ChangeEventSourceCoordinator.java [hunk 1]
```diff
@@ -58,6 +59,7 @@
 
     private volatile boolean running;
     private volatile StreamingChangeEventSource streamingSource;
+    private final ReentrantLock commitOffsetLock = new ReentrantLock();
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/pipeline/ChangeEventSourceCoordinator.java:62]: Couldn't we simply use `synchronized(streamingSource) { ... }`? Is there any advantage to using the lock?

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresSnapshotChangeEventSource.java [hunk 2]
```diff
@@ -78,10 +81,9 @@ protected SnapshotContext prepare(ChangeEventSourceContext context) throws Excep
 
     @Override
     protected void connectionCreated(RelationalSnapshotContext snapshotContext) throws Exception {
-        LOGGER.info("Setting isolation level");
-        String transactionStatement = snapshotter.snapshotTransactionIsolationLevelStatement(slotCreatedInfo);
-        LOGGER.info("Opening transaction with statement {}", transactionStatement);
-        jdbcConnection.executeWithoutCommitting(transactionStatement);
+        if (snapshotter.shouldStreamEventsStartingFromSnapshot() && startingSlotInfo == null) {
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresSnapshotChangeEventSource.java:84]: Where's the isolation level set then in case we _do_ the catch-up reading? Or asked in another way: why is it done differently for catch-up reading vs. not catch-up reading?

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresTaskContext.java [hunk 3]
```diff
@@ -66,10 +66,6 @@ protected PostgresConnectorConfig config() {
 
     protected void refreshSchema(PostgresConnection connection, boolean printReplicaIdentityInfo) throws SQLException {
         schema.refresh(connection, printReplicaIdentityInfo);
-        // Open transaction unnecessary during task execution
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresTaskContext.java:69]: Is this strictly related to the change at hand?

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java [hunk 4]
```diff
@@ -191,7 +198,9 @@ else if (message.getOperation() == Operation.COMMIT) {
                         pauseNoMessage.sleepWhen(true);
                     }
                 }
-                connection.commit();
+                if (!isInPreSnapshotCatchUpStreaming()) {
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java:201]: Why is the conditional here?

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresSnapshotChangeEventSource.java (hunk 2 above)
Comment (grantcooksey) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresSnapshotChangeEventSource.java:84]: Setting the isolation level is skipped here because it has already been set. For catch up streaming:
1. The snapshot transaction is opened and the current tx id is recorded. This transaction remains open.
2. Catch up streaming is performed until the tx id in the previous step is reached.
3. The snapshot starts. Since the snapshot transaction is already open and has executed statements, setting the isolation level is skipped.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresTaskContext.java (hunk 3 above)
Comment (grantcooksey) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresTaskContext.java:69]: Yes. We need to hold a transaction open so that we do not lose the view of data that the snapshot uses so this block must not commit during catch up streaming. I figured passing the `OffsetContext` here would be more confusing than helpful and would be duplicated by `PostgresStreamingChangeEventSource#isInPreSnapshotCatchUpStreaming`  so I moved the logic around when to commit up a level into the `PostgresStreamingChangeEventSource`.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java (hunk 4 above)
Comment (grantcooksey) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java:201]: Related to the above comments but we need to hold a transaction open through the catch up streaming so that we can stream up a specific lsn and not lose the view of data at the transaction opened at that location. Since we need to hold the transaction open for the snapshot, this block must not commit during catch up streaming.

File: debezium-core/src/main/java/io/debezium/pipeline/ChangeEventSourceCoordinator.java (hunk 1 above)
Comment (grantcooksey) [reply debezium-core/src/main/java/io/debezium/pipeline/ChangeEventSourceCoordinator.java:62]: I'm less familiar with `synchronized` so you might be right but I added the lock to protect against the `streamingSource` getting set to null between https://github.com/debezium/debezium/pull/1596/files#diff-4155443cdda33d58a9cf05816667605bR145-R146

I saw this failure pop up as a NPE during a test.

It's my understanding that `synchronized(streamingSource) { ... }` would prevent multiple threads from accessing the block for the same object but I don't see how that would prevent the race condition of different threads acting on `streamingSource` between `ChangeEventSourceCoordinator#start` and `ChangeEventSourceCoordinator#commitOffset`

Comment (grantcooksey) [reply debezium-core/src/main/java/io/debezium/pipeline/ChangeEventSourceCoordinator.java:62]: Now that I've had my ☕ , I'm thinking a little more creatively! I could see pushing the access logic into a separate function might work, ie:

```
private static void accessStreamingSource(Consumer<StreamingChangeEventSource> consumer) {
    synchronized(streamingSource) {
        consumer.accept(streamingSource);
    }
}
``` 

Is this along the lines of what you mean?

Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/pipeline/ChangeEventSourceCoordinator.java:62]: Essentially, yes. I don't think the consumer indirection would be needed, just use synchronized(streamingSource) { ... } wherever it's accessed.

Comment (grantcooksey) [reply debezium-core/src/main/java/io/debezium/pipeline/ChangeEventSourceCoordinator.java:62]: I'm a bit confused here. We really only have one thread at a time inside `start` so a synchronized block there wouldn't block another thread from entering `commitOffset` and potentially hitting a race condition, or is my understanding here off?

Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/pipeline/ChangeEventSourceCoordinator.java:62]: Not sure whether I get the question right, but essentially *every* usage to `streamingSource` must be guarded by a synchronized block; this will ensure that only ever one of these methods will be in the critical section where it accesses that source.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresSnapshotChangeEventSource.java (hunk 2 above)
Comment (gunnarmorling) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresSnapshotChangeEventSource.java:84]: Ok, thanks. That makes sense. Could you put a comment along those lines into the code?

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresTaskContext.java (hunk 3 above)
Comment (gunnarmorling) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresTaskContext.java:69]: Ok, thanks.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java (hunk 4 above)
Comment (gunnarmorling) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java:201]: Ok, gotcha. So as suggested above, having one comment somewhere in the code about the general approach for TX handling/lifecycle would be appreciated.

**Comment by gunnarmorling** [comment]
@grantcooksey, thanks a lot for the explanations, much appreciated! I think we're on the home stretch with this one, remaining issues revolve around docs and that synchronization issue.

We'll miss 1.2.0.Final with this one (as we're already in CR phase), but this will go right into the first 1.3 Alpha release due shortly after 1.2. Note for 1.3 we're planning to rework the snapshotting topic more holistically (there are a few issues related to that in the 1.3-next and 1.3-backlog Fix Versions in Jira), so this will nicely match. Also a good time to come back to your earlier PR around snapshot parallelization, and of course other things you'd consider relevant in this context.

**Comment by gunnarmorling** [comment]
Closing in favour of #1627. Thanks, @grantcooksey!

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java (hunk 4 above)
Comment (grantcooksey) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java:201]: Addressed this comment on #1627

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresSnapshotChangeEventSource.java (hunk 2 above)
Comment (grantcooksey) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresSnapshotChangeEventSource.java:84]: Addressed this comment on #1627
